File: Example/FWFramework/Classes/Module/Test/Plugin/TestRefreshController.m
//
//  TestRefreshController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestRefreshController.h"
#import "AppSwift.h"
@import FWFramework;

@interface TestRefreshObject : NSObject

@property (nonatomic, copy) NSString *title;

@property (nonatomic, copy) NSString *text;

@property (nonatomic, strong) UIImage *image;

@end

@implementation TestRefreshObject

@end

@interface TestRefreshCell : UITableViewCell

@property (nonatomic, strong) TestRefreshObject *object;

@property (nonatomic, strong) UILabel *myTitleLabel;

@property (nonatomic, strong) UILabel *myTextLabel;

@property (nonatomic, strong) UIImageView *myImageView;

@end

@implementation TestRefreshCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.fw_separatorInset = UIEdgeInsetsZero;
        
        UILabel *titleLabel = [[UILabel alloc] init];
        titleLabel.numberOfLines = 0;
        titleLabel.font = [UIFont fw_fontOfSize:15];
        titleLabel.textColor = [AppTheme textColor];
        self.myTitleLabel = titleLabel;
        [self.contentView addSubview:titleLabel]; {
            [titleLabel fw_pinEdgeToSuperview:NSLayoutAttributeLeft withInset:15];
            [titleLabel fw_pinEdgeToSuperview:NSLayoutAttributeRight withInset:15];
            NSLayoutConstraint *constraint = [titleLabel fw_pinEdgeToSuperview:NSLayoutAttributeTop withInset:15];
            [titleLabel fw_addCollapseConstraint:constraint];
            titleLabel.fw_autoCollapse = YES;
        }
        
        UILabel *textLabel = [[UILabel alloc] init];
        textLabel.numberOfLines = 0;
        textLabel.font = [UIFont fw_fontOfSize:13];
        textLabel.textColor = [AppTheme textColor];
        textLabel.fw_hiddenCollapse = YES;
        self.myTextLabel = textLabel;
        [self.contentView addSubview:textLabel]; {
            [textLabel fw_pinEdgeToSuperview:NSLayoutAttributeLeft withInset:15];
            [textLabel fw_pinEdgeToSuperview:NSLayoutAttributeRight withInset:15];
            NSLayoutConstraint *constraint = [textLabel fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:titleLabel withOffset:10];
            [textLabel fw_addCollapseConstraint:constraint];
        }
        
        UIImageView *imageView = [[UIImageView alloc] init];
        self.myImageView = imageView;
        [self.contentView addSubview:imageView]; {
            [imageView fw_pinEdgeToSuperview:NSLayoutAttributeLeft withInset:15];
            [imageView fw_pinEdgeToSuperview:NSLayoutAttributeRight withInset:15 relation:NSLayoutRelationGreaterThanOrEqual];
            [imageView fw_pinEdgeToSuperview:NSLayoutAttributeBottom withInset:15];
            NSLayoutConstraint *constraint = [imageView fw_pinEdge:NSLayoutAttributeTop toEdge:NSLayoutAttributeBottom ofView:textLabel withOffset:10];
            [imageView fw_addCollapseConstraint:constraint];
            imageView.fw_autoCollapse = YES;
        }
    }
    return self;
}

- (void)setObject:(TestRefreshObject *)object
{
    _object = object;
    // 自动收缩
    self.myTitleLabel.text = object.title;
    self.myImageView.image = object.image;
    // 隐藏时自动收缩
    self.myTextLabel.text = object.text;
    self.myTextLabel.hidden = object.text.length < 1;
}

@end

@interface TestRefreshController () <FWTableViewController>

@end

@implementation TestRefreshController

- (UITableViewStyle)setupTableStyle
{
    return UITableViewStyleGrouped;
}

- (void)setupTableView
{
    [self.tableView fw_resetGroupedStyle];
    [self.tableView registerClass:[TestRefreshCell class] forCellReuseIdentifier:@"Cell"];
}

- (void)setupSubviews
{
    FWInfiniteScrollView.height = 64;
    [self.tableView fw_setRefreshingTarget:self action:@selector(onRefreshing)];
    [self.tableView fw_setLoadingTarget:self action:@selector(onLoading)];
    
    UIImageView *pullView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
    pullView.image = [FWModuleBundle imageNamed:@"Loading.gif"];
    self.tableView.fw_pullRefreshView.shouldChangeAlpha = NO;
    [self.tableView.fw_pullRefreshView setCustomView:pullView forState:FWPullRefreshStateAll];
    FWWeakifySelf();
    self.tableView.fw_pullRefreshView.stateBlock = ^(FWPullRefreshView * _Nonnull view, FWPullRefreshState state) {
        FWStrongifySelf();
        self.navigationItem.title = [NSString stringWithFormat:@"refresh state-%@", @(state)];
    };
    self.tableView.fw_pullRefreshView.progressBlock = ^(FWPullRefreshView * _Nonnull view, CGFloat progress) {
        FWStrongifySelf();
        self.navigationItem.title = [NSString stringWithFormat:@"refresh progress-%.2f", progress];
    };
    
    UIImageView *infiniteView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
    infiniteView.image = [FWModuleBundle imageNamed:@"Loading.gif"];
    [self.tableView.fw_infiniteScrollView setCustomView:infiniteView forState:FWInfiniteScrollStateAll];
    self.tableView.fw_infiniteScrollView.stateBlock = ^(FWInfiniteScrollView * _Nonnull view, FWInfiniteScrollState state) {
        FWStrongifySelf();
        self.navigationItem.title = [NSString stringWithFormat:@"load state-%@", @(state)];
    };
    self.tableView.fw_infiniteScrollView.progressBlock = ^(FWInfiniteScrollView * _Nonnull view, CGFloat progress) {
        FWStrongifySelf();
        self.navigationItem.title = [NSString stringWithFormat:@"load progress-%.2f", progress];
    };
}

- (void)setupLayout
{
    [self.tableView fw_beginLoading];
}

#pragma mark - TableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.tableData.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    // 渲染可重用Cell
    TestRefreshCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    TestRefreshObject *object = [self.tableData objectAtIndex:indexPath.row];
    cell.object = object;
    return cell;
}

- (TestRefreshObject *)randomObject
{
    static NSMutableArray<NSArray *> *randomArray;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        randomArray = [NSMutableArray array];
        
        [randomArray addObject:@[
            @"",
            @"这是标题",
            @"这是复杂的标题这是复杂的标题这是复杂的标题",
            @"这是复杂的标题这是复杂的标题\n这是复杂的标题这是复杂的标题",
            @"这是复杂的标题\n这是复杂的标题\n这是复杂的标题\n这是复杂的标题",
        ]];
        
        [randomArray addObject:@[
            @"",
            @"这是内容",
            @"这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容",
            @"这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容",
            @"这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容",
        ]];
        
        [randomArray addObject:@[
            @"",
            @"Animation.png",
        ]];
    });
    
    TestRefreshObject *object = [TestRefreshObject new];
    object.title = [[randomArray objectAtIndex:0] fw_randomObject];
    object.text = [[randomArray objectAtIndex:1] fw_randomObject];
    NSString *imageName =[[randomArray objectAtIndex:2] fw_randomObject];
    if (imageName.length > 0) {
        object.image = [FWModuleBundle imageNamed:imageName];
    }
    return object;
}

- (void)onRefreshing
{
    NSLog(@"开始刷新");
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        NSLog(@"刷新完成");
        
        [self.tableData removeAllObjects];
        for (int i = 0; i < 1; i++) {
            [self.tableData addObject:[self randomObject]];
        }
        [self.tableView reloadData];
        
        self.tableView.fw_shouldRefreshing = self.tableData.count < 20 ? YES : NO;
        [self.tableView fw_endRefreshing];
    });
}

- (void)onLoading
{
    NSLog(@"开始加载");
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        NSLog(@"加载完成");
        
        for (int i = 0; i < 1; i++) {
            [self.tableData addObject:[self randomObject]];
        }
        [self.tableView reloadData];
        
        self.tableView.fw_loadingFinished = self.tableData.count >= 20 ? YES : NO;
        [self.tableView fw_endLoading];
    });
}

@end

File: Example/FWFramework/Classes/Module/Test/Plugin/TestRefreshController.swift
//
//  TestRefreshController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/19.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestRefreshController: UIViewController, TableViewControllerProtocol, EmptyViewDelegate {
    
    var emptyViewMode: Int = 0
    
    func setupTableStyle() -> UITableView.Style {
        .grouped
    }
    
    func setupTableView() {
        tableView.app.resetTableStyle()
        tableView.contentInsetAdjustmentBehavior = .never
        tableView.alwaysBounceVertical = true
        tableView.register(TestRefreshCell.self, forCellReuseIdentifier: "Cell")
    }
    
    func setupNavbar() {
        app.extendedLayoutEdge = []
        app.setRightBarItem(UIBarButtonItem.SystemItem.action.rawValue) { [weak self] _ in
            self?.app.showSheet(title: nil, message: nil, actions: [self?.emptyViewMode == 1 ? "隐藏view空界面" : "显示view空界面", self?.emptyViewMode == 2 ? "隐藏tableView空界面" : "显示tableView空界面"], actionBlock: { index in
                if index == 0 {
                    self?.emptyViewMode = self?.emptyViewMode == 1 ? 0 : 1
                } else {
                    self?.emptyViewMode = self?.emptyViewMode == 2 ? 0 : 2
                }
                self?.tableView.app.beginRefreshing()
            })
        }
    }
    
    func setupSubviews() {
        // InfiniteScrollView.height = 64
        tableView.app.setRefreshing(target: self, action: #selector(onRefreshing))
        tableView.app.setLoading(target: self, action: #selector(onLoading))
        
        // let pullView = UIImageView(frame: CGRect(x: 0, y: 0, width: 40, height: 40))
        // pullView.image = ModuleBundle.imageNamed("Loading.gif")
        // tableView.app.pullRefreshView?.shouldChangeAlpha = false
        // tableView.app.pullRefreshView?.setCustomView(pullView, for: .all)
        tableView.app.pullRefreshView?.stateBlock = { [weak self] view, state in
            self?.navigationItem.title = "refresh state-\(state.rawValue)"
        }
        tableView.app.pullRefreshView?.progressBlock = { [weak self] view, progress in
            self?.navigationItem.title = String(format: "refresh progress-%.2f", progress)
        }
        
        // let infiniteView = UIImageView(frame: CGRect(x: 0, y: 0, width: 40, height: 40))
        // infiniteView.image = ModuleBundle.imageNamed("Loading.gif")
        // tableView.app.infiniteScrollView?.setCustomView(infiniteView, for: .all)
        // tableView.app.infiniteScrollView?.preloadHeight = 200
        tableView.app.infiniteScrollView?.stateBlock = { [weak self] view, state in
            self?.navigationItem.title = "load state-\(state.rawValue)"
        }
        tableView.app.infiniteScrollView?.progressBlock = { [weak self] view, progress in
            self?.navigationItem.title = String(format: "load progress-%.2f", progress)
        }
    }
    
    func setupLayout() {
        tableView.app.beginRefreshing()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return tableData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath) as! TestRefreshCell
        cell.object = tableData[indexPath.row] as? TestRefreshObject
        return cell
    }
    
    func emptyViewShouldScroll(_ scrollView: UIScrollView) -> Bool {
        return true
    }
    
    func randomObject() -> TestRefreshObject {
        let object = TestRefreshObject()
        object.title = [
            "",
            "这是标题",
            "这是复杂的标题这是复杂的标题这是复杂的标题",
            "这是复杂的标题这是复杂的标题\n这是复杂的标题这是复杂的标题",
            "这是复杂的标题\n这是复杂的标题\n这是复杂的标题\n这是复杂的标题",
        ].randomElement() ?? ""
        object.text = [
            "",
            "这是内容",
            "这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容这是复杂的内容",
            "这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容",
            "这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容\n这是复杂的内容这是复杂的内容",
        ].randomElement() ?? ""
        let imageName = [
            "",
            "Animation.png",
        ].randomElement() ?? ""
        if !imageName.isEmpty {
            object.image = ModuleBundle.imageNamed(imageName)
        }
        return object
    }
    
    @objc func onRefreshing() {
        NSLog("开始刷新")
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
            NSLog("刷新完成")
            
            self.view.app.hideEmptyView()
            self.tableView.app.hideEmptyView()
            self.tableView.app.emptyViewDelegate = nil
            
            if self.emptyViewMode == 0 {
                self.tableData.removeAll()
                for _ in 0 ..< 5 {
                    self.tableData.append(self.randomObject())
                }
                self.tableView.reloadData()
                self.tableView.app.endRefreshing(finished: self.tableData.count >= 10)
            } else if self.emptyViewMode == 1 {
                self.tableData.removeAll()
                self.tableView.reloadData()
                self.view.app.showEmptyView()
                self.tableView.app.endRefreshing(finished: true)
            } else {
                self.tableData.removeAll()
                self.tableView.app.emptyViewDelegate = self
                self.tableView.reloadData()
                self.tableView.app.endRefreshing(finished: true)
            }
        }
    }
    
    @objc func onLoading() {
        NSLog("开始加载")
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
            NSLog("加载完成")
            
            for _ in 0 ..< 5 {
                self.tableData.append(self.randomObject())
            }
            self.tableView.reloadData()
            self.tableView.app.endLoading(finished: self.tableData.count >= 10)
        }
    }
    
}

class TestRefreshObject: NSObject {
    
    var title: String = ""
    var text: String = ""
    var image: UIImage?
    
}

class TestRefreshCell: UITableViewCell {
    
    var object: TestRefreshObject? {
        didSet {
            myTitleLabel.text = object?.title
            myImageView.image = object?.image
            myTextLabel.text = object?.text
            myTextLabel.isHidden = APP.safeString(object?.text).isEmpty
        }
    }
    
    lazy var myTitleLabel: UILabel = {
        let result = UILabel()
        result.numberOfLines = 0
        result.font = UIFont.app.font(ofSize: 15)
        result.textColor = AppTheme.textColor
        return result
    }()
    
    lazy var myTextLabel: UILabel = {
        let result = UILabel()
        result.numberOfLines = 0
        result.font = UIFont.app.font(ofSize: 13)
        result.textColor = AppTheme.textColor
        return result
    }()
    
    lazy var myImageView: UIImageView = {
        let result = UIImageView()
        return result
    }()
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        app.separatorInset = .zero
        contentView.addSubview(myTitleLabel)
        contentView.addSubview(myTextLabel)
        contentView.addSubview(myImageView)
        
        myTitleLabel.chain
            .horizontal(15)
            .top(15).collapse()
            .autoCollapse(true)
        
        myTextLabel.chain
            .hiddenCollapse(true)
            .horizontal(15)
            .top(toViewBottom: myTitleLabel, offset: 10).collapse()
        
        myImageView.chain
            .left(15)
            .right(15, relation: .greaterThanOrEqual)
            .bottom(15)
            .top(toViewBottom: myTextLabel, offset: 10).collapse()
            .autoCollapse(true)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
